File: sample.js
/**
 * Aplikasi Kompres & De-kompres
 * =============================
 * 
 * Aplikasi web modern untuk kompresi file gambar, video, dan kode.
 * Mendukung berbagai format file dengan pengaturan kompresi yang fleksibel.
 * 
 * @author AI Assistant
 * @version 2.0.0
 * @license MIT
 */

'use strict';

/**
 * Enum untuk jenis kompresi yang didukung
 */
const CompressionType = {
    IMAGE: 'image',
    VIDEO: 'video',
    CODE: 'code',
    FILE: 'file'
};

/**
 * Enum untuk format gambar yang didukung
 */
const ImageFormat = {
    JPEG: 'jpeg',
    PNG: 'png',
    WEBP: 'webp',
    GIF: 'gif'
};

/**
 * Enum untuk format video yang didukung
 */
const VideoFormat = {
    MP4: 'mp4',
    WEBM: 'webm',
    AVI: 'avi',
    MOV: 'mov'
};

/**
 * Class untuk pengaturan kompresi
 */
class CompressionSettings {
    constructor(options = {}) {
        this.quality = options.quality || 80;
        this.maxWidth = options.maxWidth || 1920;
        this.maxHeight = options.maxHeight || 1080;
        this.fps = options.fps || 30;
        this.bitrate = options.bitrate || null;
        this.removeComments = options.removeComments !== false;
        this.removeWhitespace = options.removeWhitespace !== false;
        this.minifyKeys = options.minifyKeys || false;
    }
}

/**
 * Kelas utama untuk menangani kompresi berbagai jenis file
 */
class FileCompressor {
    constructor() {
        this.supportedImageFormats = ['.jpg', '.jpeg', '.png', '.webp', '.gif'];
        this.supportedVideoFormats = ['.mp4', '.webm', '.avi', '.mov'];
        this.supportedCodeFormats = ['.js', '.html', '.css', '.json', '.xml', '.py', '.php'];
        
        this.defaultSettings = new CompressionSettings();
        
        console.log('FileCompressor initialized successfully');
    }
    
    /**
     * Mendeteksi jenis file berdasarkan ekstensi
     * 
     * @param {string} filePath - Path ke file yang akan dianalisis
     * @returns {string} Jenis kompresi yang sesuai
     * @throws {Error} Jika format file tidak didukung
     */
    detectFileType(filePath) {
        const fileExtension = '.' + filePath.split('.').pop().toLowerCase();
        
        if (this.supportedImageFormats.includes(fileExtension)) {
            return CompressionType.IMAGE;
        } else if (this.supportedVideoFormats.includes(fileExtension)) {
            return CompressionType.VIDEO;
        } else if (this.supportedCodeFormats.includes(fileExtension)) {
            return CompressionType.CODE;
        } else {
            return CompressionType.FILE;
        }
    }
    
    /**
     * Kompresi file gambar dengan pengaturan yang diberikan
     * 
     * @param {string} inputPath - Path file input
     * @param {string} outputPath - Path file output
     * @param {CompressionSettings} settings - Pengaturan kompresi
     * @returns {Promise<Object>} Statistik kompresi
     * @throws {Error} Jika terjadi error saat kompresi
     */
    async compressImage(inputPath, outputPath, settings) {
        try {
            // Simulasi kompresi gambar
            const originalSize = 1024 * 1024; // 1MB simulasi
            const compressedSize = Math.floor(originalSize * (settings.quality / 100));
            
            // Simulasi penulisan file terkompresi
            const compressedContent = `Compressed image with quality ${settings.quality}%`;
            
            const compressionRatio = ((originalSize - compressedSize) / originalSize) * 100;
            
            console.log(`Image compressed successfully: ${compressionRatio.toFixed(2)}% reduction`);
            
            return {
                originalSize: originalSize,
                compressedSize: compressedSize,
                compressionRatio: compressionRatio,
                quality: settings.quality
            };
            
        } catch (error) {
            console.error('Error compressing image:', error.message);
            throw new Error(`Failed to compress image: ${error.message}`);
        }
    }
    
    /**
     * Kompresi file video dengan pengaturan yang diberikan
     * 
     * @param {string} inputPath - Path file input
     * @param {string} outputPath - Path file output
     * @param {CompressionSettings} settings - Pengaturan kompresi
     * @returns {Promise<Object>} Statistik kompresi
     * @throws {Error} Jika terjadi error saat kompresi
     */
    async compressVideo(inputPath, outputPath, settings) {
        try {
            // Simulasi kompresi video
            const originalSize = 50 * 1024 * 1024; // 50MB simulasi
            const compressedSize = Math.floor(originalSize * (settings.quality / 100) * 0.8);
            
            // Simulasi penulisan file terkompresi
            const compressedContent = `Compressed video with quality ${settings.quality}% and ${settings.fps} FPS`;
            
            const compressionRatio = ((originalSize - compressedSize) / originalSize) * 100;
            
            console.log(`Video compressed successfully: ${compressionRatio.toFixed(2)}% reduction`);
            
            return {
                originalSize: originalSize,
                compressedSize: compressedSize,
                compressionRatio: compressionRatio,
                quality: settings.quality,
                fps: settings.fps
            };
            
        } catch (error) {
            console.error('Error compressing video:', error.message);
            throw new Error(`Failed to compress video: ${error.message}`);
        }
    }
    
    /**
     * Kompresi file kode dengan pengaturan yang diberikan
     * 
     * @param {string} inputPath - Path file input
     * @param {string} outputPath - Path file output
     * @param {CompressionSettings} settings - Pengaturan kompresi
     * @returns {Promise<Object>} Statistik kompresi
     * @throws {Error} Jika terjadi error saat kompresi
     */
    async compressCode(inputPath, outputPath, settings) {
        try {
            // Simulasi membaca file
            const originalCode = this.getSampleCode();
            let compressedCode = originalCode;
            
            // Hapus komentar jika diaktifkan
            if (settings.removeComments) {
                compressedCode = this.removeComments(compressedCode);
            }
            
            // Hapus whitespace jika diaktifkan
            if (settings.removeWhitespace) {
                compressedCode = this.removeWhitespace(compressedCode);
            }
            
            // Minify keys jika diaktifkan
            if (settings.minifyKeys) {
                compressedCode = this.minifyKeys(compressedCode);
            }
            
            const originalSize = originalCode.length;
            const compressedSize = compressedCode.length;
            const compressionRatio = ((originalSize - compressedSize) / originalSize) * 100;
            
            console.log(`Code compressed successfully: ${compressionRatio.toFixed(2)}% reduction`);
            
            return {
                originalSize: originalSize,
                compressedSize: compressedSize,
                compressionRatio: compressionRatio,
                charactersRemoved: originalSize - compressedSize
            };
            
        } catch (error) {
            console.error('Error compressing code:', error.message);
            throw new Error(`Failed to compress code: ${error.message}`);
        }
    }
    
    /**
     * Hapus komentar dari kode JavaScript
     * 
     * @param {string} code - Kode yang akan dibersihkan
     * @returns {string} Kode tanpa komentar
     */
    removeComments(code) {
        // Hapus komentar multi-line /* */
        code = code.replace(/\/\*[\s\S]*?\*\//g, '');
        
        // Hapus komentar single-line //
        code = code.replace(/\/\/.*$/gm, '');
        
        return code;
    }
    
    /**
     * Hapus whitespace berlebih dari kode
     * 
     * @param {string} code - Kode yang akan dibersihkan
     * @returns {string} Kode tanpa whitespace berlebih
     */
    removeWhitespace(code) {
        // Hapus spasi di awal dan akhir
        code = code.trim();
        
        // Hapus baris kosong berlebih
        const lines = code.split('\n');
        const cleanedLines = [];
        
        for (const line of lines) {
            const trimmedLine = line.trim();
            if (trimmedLine) {
                cleanedLines.push(trimmedLine);
            }
        }
        
        return cleanedLines.join('\n');
    }
    
    /**
     * Minify key names dalam JSON atau object
     * 
     * @param {string} code - Kode yang akan diminify
     * @returns {string} Kode dengan key yang diminify
     */
    minifyKeys(code) {
        // Implementasi sederhana untuk minify keys
        code = code.replace(/"name"/g, '"n"');
        code = code.replace(/"description"/g, '"d"');
        code = code.replace(/"title"/g, '"t"');
        code = code.replace(/"content"/g, '"c"');
        code = code.replace(/"value"/g, '"v"');
        
        return code;
    }
    
    /**
     * Mendapatkan statistik file sebelum kompresi
     * 
     * @param {string} inputPath - Path ke file input
     * @returns {Object} Statistik file
     * @throws {Error} Jika terjadi error saat membaca file
     */
    getCompressionStats(inputPath) {
        try {
            const fileType = this.detectFileType(inputPath);
            const fileSize = 1024 * 1024; // 1MB simulasi
            
            const stats = {
                filePath: inputPath,
                fileSize: fileSize,
                fileType: fileType,
                fileExtension: '.' + inputPath.split('.').pop().toLowerCase(),
                canCompress: fileType !== CompressionType.FILE
            };
            
            // Tambahan statistik berdasarkan jenis file
            if (fileType === CompressionType.CODE) {
                const sampleCode = this.getSampleCode();
                stats.lineCount = sampleCode.split('\n').length;
                stats.characterCount = sampleCode.length;
                stats.wordCount = sampleCode.split(/\s+/).length;
            }
            
            return stats;
            
        } catch (error) {
            console.error('Error getting file stats:', error.message);
            throw new Error(`Failed to get file stats: ${error.message}`);
        }
    }
    
    /**
     * Mendapatkan daftar format file yang didukung
     * 
     * @returns {Object} Daftar format file yang didukung
     */
    getSupportedFormats() {
        return {
            images: this.supportedImageFormats,
            videos: this.supportedVideoFormats,
            code: this.supportedCodeFormats
        };
    }
    
    /**
     * Mendapatkan sample code untuk testing
     * 
     * @returns {string} Sample code
     */
    getSampleCode() {
        return `
            // Sample JavaScript code for testing
            const sampleData = {
                name: "Test User",
                description: "This is a test user for compression testing",
                age: 25,
                email: "test@example.com",
                preferences: {
                    theme: "dark",
                    language: "en",
                    notifications: true
                }
            };
            
            function processData(data) {
                // Process the data
                const processed = {
                    ...data,
                    processed: true,
                    timestamp: new Date().toISOString()
                };
                
                return processed;
            }
            
            // Export the function
            module.exports = {
                processData,
                sampleData
            };
        `;
    }
}

/**
 * Utility functions
 */
class CompressionUtils {
    /**
     * Format bytes ke human readable format
     * 
     * @param {number} bytes - Jumlah bytes
     * @param {number} decimals - Jumlah desimal
     * @returns {string} Formatted string
     */
    static formatBytes(bytes, decimals = 2) {
        if (bytes === 0) return '0 Bytes';
        
        const k = 1024;
        const dm = decimals < 0 ? 0 : decimals;
        const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB'];
        
        const i = Math.floor(Math.log(bytes) / Math.log(k));
        
        return parseFloat((bytes / Math.pow(k, i)).toFixed(dm)) + ' ' + sizes[i];
    }
    
    /**
     * Format duration ke human readable format
     * 
     * @param {number} seconds - Jumlah detik
     * @returns {string} Formatted string
     */
    static formatDuration(seconds) {
        const hours = Math.floor(seconds / 3600);
        const minutes = Math.floor((seconds % 3600) / 60);
        const secs = Math.floor(seconds % 60);
        
        if (hours > 0) {
            return `${hours}:${minutes.toString().padStart(2, '0')}:${secs.toString().padStart(2, '0')}`;
        } else {
            return `${minutes}:${secs.toString().padStart(2, '0')}`;
        }
    }
    
    /**
     * Format bitrate ke human readable format
     * 
     * @param {number} bitsPerSecond - Bitrate dalam bits per second
     * @returns {string} Formatted string
     */
    static formatBitrate(bitsPerSecond) {
        if (bitsPerSecond >= 1000000000) {
            return (bitsPerSecond / 1000000000).toFixed(1) + ' Gbps';
        } else if (bitsPerSecond >= 1000000) {
            return (bitsPerSecond / 1000000).toFixed(1) + ' Mbps';
        } else if (bitsPerSecond >= 1000) {
            return (bitsPerSecond / 1000).toFixed(1) + ' Kbps';
        } else {
            return bitsPerSecond + ' bps';
        }
    }
}

/**
 * Fungsi utama untuk demonstrasi
 */
function main() {
    console.log('=== Aplikasi Kompres & De-kompres ===');
    console.log('Versi: 2.0.0');
    console.log('Author: AI Assistant');
    console.log('Language: JavaScript');
    console.log();
    
    // Inisialisasi compressor
    const compressor = new FileCompressor();
    
    // Contoh penggunaan
    const testFiles = [
        'sample.json',
        'sample.py',
        'sample.php',
        'sample.js'
    ];
    
    for (const filePath of testFiles) {
        try {
            const stats = compressor.getCompressionStats(filePath);
            console.log(`File: ${filePath}`);
            console.log(`  Size: ${CompressionUtils.formatBytes(stats.fileSize)}`);
            console.log(`  Type: ${stats.fileType}`);
            console.log(`  Can compress: ${stats.canCompress ? 'Yes' : 'No'}`);
            
            if (stats.fileType === CompressionType.CODE) {
                console.log(`  Lines: ${stats.lineCount}`);
                console.log(`  Characters: ${stats.characterCount}`);
                console.log(`  Words: ${stats.wordCount}`);
            }
            
            console.log();
            
        } catch (error) {
            console.error(`Error processing ${filePath}:`, error.message);
        }
    }
    
    console.log('Demo completed successfully!');
}

// Export classes untuk penggunaan di module lain
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        FileCompressor,
        CompressionSettings,
        CompressionType,
        ImageFormat,
        VideoFormat,
        CompressionUtils,
        main
    };
}

// Jalankan fungsi utama jika dijalankan langsung
if (typeof window === 'undefined') {
    main();
} 

File: sample.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Aplikasi Kompres & De-kompres
=============================

Aplikasi web modern untuk kompresi file gambar, video, dan kode.
Mendukung berbagai format file dengan pengaturan kompresi yang fleksibel.

Author: AI Assistant
Version: 2.0.0
License: MIT
"""

import json
import os
import sys
from typing import Dict, List, Optional, Tuple, Union
from dataclasses import dataclass
from enum import Enum
import logging
from pathlib import Path

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


class CompressionType(Enum):
    """Enum untuk jenis kompresi yang didukung."""
    IMAGE = "image"
    VIDEO = "video"
    CODE = "code"
    FILE = "file"


class ImageFormat(Enum):
    """Enum untuk format gambar yang didukung."""
    JPEG = "jpeg"
    PNG = "png"
    WEBP = "webp"
    GIF = "gif"


class VideoFormat(Enum):
    """Enum untuk format video yang didukung."""
    MP4 = "mp4"
    WEBM = "webm"
    AVI = "avi"
    MOV = "mov"


@dataclass
class CompressionSettings:
    """Data class untuk pengaturan kompresi."""
    quality: int = 80
    max_width: int = 1920
    max_height: int = 1080
    fps: int = 30
    bitrate: Optional[int] = None
    remove_comments: bool = True
    remove_whitespace: bool = True
    minify_keys: bool = False


class FileCompressor:
    """
    Kelas utama untuk menangani kompresi berbagai jenis file.
    
    Attributes:
        supported_image_formats (List[str]): Format gambar yang didukung
        supported_video_formats (List[str]): Format video yang didukung
        supported_code_formats (List[str]): Format kode yang didukung
    """
    
    def __init__(self):
        self.supported_image_formats = ['.jpg', '.jpeg', '.png', '.webp', '.gif']
        self.supported_video_formats = ['.mp4', '.webm', '.avi', '.mov']
        self.supported_code_formats = ['.py', '.js', '.php', '.html', '.css', '.json', '.xml']
        
        # Initialize compression settings
        self.default_settings = CompressionSettings()
        
        logger.info("FileCompressor initialized successfully")
    
    def detect_file_type(self, file_path: str) -> CompressionType:
        """
        Mendeteksi jenis file berdasarkan ekstensi.
        
        Args:
            file_path (str): Path ke file yang akan dianalisis
            
        Returns:
            CompressionType: Jenis kompresi yang sesuai
            
        Raises:
            ValueError: Jika format file tidak didukung
        """
        file_extension = Path(file_path).suffix.lower()
        
        if file_extension in self.supported_image_formats:
            return CompressionType.IMAGE
        elif file_extension in self.supported_video_formats:
            return CompressionType.VIDEO
        elif file_extension in self.supported_code_formats:
            return CompressionType.CODE
        else:
            return CompressionType.FILE
    
    def compress_image(self, 
                      input_path: str, 
                      output_path: str, 
                      settings: CompressionSettings) -> Dict[str, Union[int, float]]:
        """
        Kompresi file gambar dengan pengaturan yang diberikan.
        
        Args:
            input_path (str): Path file input
            output_path (str): Path file output
            settings (CompressionSettings): Pengaturan kompresi
            
        Returns:
            Dict[str, Union[int, float]]: Statistik kompresi
        """
        try:
            # Simulasi kompresi gambar
            original_size = os.path.getsize(input_path)
            compressed_size = int(original_size * (settings.quality / 100))
            
            # Simulasi penulisan file terkompresi
            with open(output_path, 'w') as f:
                f.write(f"Compressed image with quality {settings.quality}%")
            
            compression_ratio = ((original_size - compressed_size) / original_size) * 100
            
            logger.info(f"Image compressed successfully: {compression_ratio:.2f}% reduction")
            
            return {
                'original_size': original_size,
                'compressed_size': compressed_size,
                'compression_ratio': compression_ratio,
                'quality': settings.quality
            }
            
        except Exception as e:
            logger.error(f"Error compressing image: {str(e)}")
            raise
    
    def compress_video(self, 
                      input_path: str, 
                      output_path: str, 
                      settings: CompressionSettings) -> Dict[str, Union[int, float]]:
        """
        Kompresi file video dengan pengaturan yang diberikan.
        
        Args:
            input_path (str): Path file input
            output_path (str): Path file output
            settings (CompressionSettings): Pengaturan kompresi
            
        Returns:
            Dict[str, Union[int, float]]: Statistik kompresi
        """
        try:
            # Simulasi kompresi video
            original_size = os.path.getsize(input_path)
            compressed_size = int(original_size * (settings.quality / 100) * 0.8)
            
            # Simulasi penulisan file terkompresi
            with open(output_path, 'w') as f:
                f.write(f"Compressed video with quality {settings.quality}% and {settings.fps} FPS")
            
            compression_ratio = ((original_size - compressed_size) / original_size) * 100
            
            logger.info(f"Video compressed successfully: {compression_ratio:.2f}% reduction")
            
            return {
                'original_size': original_size,
                'compressed_size': compressed_size,
                'compression_ratio': compression_ratio,
                'quality': settings.quality,
                'fps': settings.fps
            }
            
        except Exception as e:
            logger.error(f"Error compressing video: {str(e)}")
            raise
    
    def compress_code(self, 
                     input_path: str, 
                     output_path: str, 
                     settings: CompressionSettings) -> Dict[str, Union[int, float]]:
        """
        Kompresi file kode dengan pengaturan yang diberikan.
        
        Args:
            input_path (str): Path file input
            output_path (str): Path file output
            settings (CompressionSettings): Pengaturan kompresi
            
        Returns:
            Dict[str, Union[int, float]]: Statistik kompresi
        """
        try:
            with open(input_path, 'r', encoding='utf-8') as f:
                original_code = f.read()
            
            compressed_code = original_code
            
            # Hapus komentar jika diaktifkan
            if settings.remove_comments:
                compressed_code = self._remove_comments(compressed_code)
            
            # Hapus whitespace jika diaktifkan
            if settings.remove_whitespace:
                compressed_code = self._remove_whitespace(compressed_code)
            
            # Minify keys jika diaktifkan
            if settings.minify_keys:
                compressed_code = self._minify_keys(compressed_code)
            
            # Tulis file terkompresi
            with open(output_path, 'w', encoding='utf-8') as f:
                f.write(compressed_code)
            
            original_size = len(original_code.encode('utf-8'))
            compressed_size = len(compressed_code.encode('utf-8'))
            compression_ratio = ((original_size - compressed_size) / original_size) * 100
            
            logger.info(f"Code compressed successfully: {compression_ratio:.2f}% reduction")
            
            return {
                'original_size': original_size,
                'compressed_size': compressed_size,
                'compression_ratio': compression_ratio,
                'characters_removed': len(original_code) - len(compressed_code)
            }
            
        except Exception as e:
            logger.error(f"Error compressing code: {str(e)}")
            raise
    
    def _remove_comments(self, code: str) -> str:
        """Hapus komentar dari kode Python."""
        lines = code.split('\n')
        cleaned_lines = []
        
        for line in lines:
            # Hapus komentar inline
            comment_pos = line.find('#')
            if comment_pos != -1:
                line = line[:comment_pos].rstrip()
            
            # Skip baris kosong setelah penghapusan komentar
            if line.strip():
                cleaned_lines.append(line)
        
        return '\n'.join(cleaned_lines)
    
    def _remove_whitespace(self, code: str) -> str:
        """Hapus whitespace berlebih dari kode."""
        # Hapus spasi di awal dan akhir
        code = code.strip()
        
        # Hapus baris kosong berlebih
        lines = code.split('\n')
        cleaned_lines = []
        
        for line in lines:
            if line.strip():
                cleaned_lines.append(line.strip())
        
        return '\n'.join(cleaned_lines)
    
    def _minify_keys(self, code: str) -> str:
        """Minify key names dalam JSON atau dictionary."""
        # Implementasi sederhana untuk minify keys
        # Dalam implementasi nyata, ini akan lebih kompleks
        return code.replace('"name"', '"n"').replace('"description"', '"d"')
    
    def get_compression_stats(self, input_path: str) -> Dict[str, any]:
        """
        Mendapatkan statistik file sebelum kompresi.
        
        Args:
            input_path (str): Path ke file input
            
        Returns:
            Dict[str, any]: Statistik file
        """
        try:
            file_size = os.path.getsize(input_path)
            file_type = self.detect_file_type(input_path)
            
            stats = {
                'file_path': input_path,
                'file_size': file_size,
                'file_type': file_type.value,
                'file_extension': Path(input_path).suffix.lower(),
                'can_compress': file_type != CompressionType.FILE
            }
            
            # Tambahan statistik berdasarkan jenis file
            if file_type == CompressionType.CODE:
                with open(input_path, 'r', encoding='utf-8') as f:
                    content = f.read()
                    stats['line_count'] = len(content.split('\n'))
                    stats['character_count'] = len(content)
                    stats['word_count'] = len(content.split())
            
            return stats
            
        except Exception as e:
            logger.error(f"Error getting file stats: {str(e)}")
            raise


def main():
    """Fungsi utama untuk demonstrasi."""
    print("=== Aplikasi Kompres & De-kompres ===")
    print("Versi: 2.0.0")
    print("Author: AI Assistant")
    print()
    
    # Inisialisasi compressor
    compressor = FileCompressor()
    
    # Contoh penggunaan
    test_files = [
        "sample.json",
        "sample.py", 
        "sample.php"
    ]
    
    for file_path in test_files:
        if os.path.exists(file_path):
            try:
                stats = compressor.get_compression_stats(file_path)
                print(f"File: {file_path}")
                print(f"  Size: {stats['file_size']} bytes")
                print(f"  Type: {stats['file_type']}")
                print(f"  Can compress: {stats['can_compress']}")
                
                if stats['file_type'] == 'code':
                    print(f"  Lines: {stats.get('line_count', 'N/A')}")
                    print(f"  Characters: {stats.get('character_count', 'N/A')}")
                
                print()
                
            except Exception as e:
                print(f"Error processing {file_path}: {str(e)}")
        else:
            print(f"File {file_path} not found")
    
    print("Demo completed successfully!")


if __name__ == "__main__":
    main() 